intrinsic_and_extrinsic_optimise_real_data/src/camera_calibration_real_data.cpp

```diff
@@ -0,0 +1,422 @@
+static const double initial_fx = 950;
+static const double initial_fy = 950;
+static const double initial_cx = 320;
+static const double initial_cy = 240;
+
+// Data Structure definition
+struct ReProjectionResidual
+{ 
+	// Put the actual parameters that will not be optimised right here (this would be the intrinsics,the uv pixel points and board_T_cam).
+	ReProjectionResidual(const double *pixel_points, const double *board_T_point)
+	{
+		// Initialise the pixel points
+		observed_pixel_points_[0] = pixel_points[0]; // u
+		observed_pixel_points_[1] = pixel_points[1]; // v
+
+		// Initialise the board_T_point
+		board_T_point_[0] = board_T_point[0]; // X
+		board_T_point_[1] = board_T_point[1]; // Y
+		board_T_point_[2] = board_T_point[2]; // Z
+	}
+
+	// For this template, put all of the parameters to optimise and the residual here.
+	template <typename T>
+	bool operator()(const T* const camera_extrinsics, const T* const camera_intrinsics, T* residuals)
+	const
+	{
+		// compute projective coordinates: p = RX + t.
+        // camera_extrinsics[0, 1, 2]: axis-angle
+        // camera_extrinsics[3, 4, 5]: translation
+		const T R_angle_axis[3] = {T(camera_extrinsics[0]), T(camera_extrinsics[1]), T(camera_extrinsics[2])};
+		const T point[3] = {T(board_T_point_[0]), T(board_T_point_[1]), T(board_T_point_[2])};
+		T p[3];
+
+		// AngleAxisRotatePoint used to rotate the board_T_point about the axis of rotation which is set 
+		// as the R component of the camera extrinsic matric after the rotation matrix to angle axis conversion
+		ceres::AngleAxisRotatePoint(R_angle_axis, point, p);
+
+		// AngleAxisRotatePoint gives the "RX" therefore it must be translated by "t" (from camera extrinsics) to give "p".
+		p[0] += camera_extrinsics[3]; // X component of camera to calibration point
+		p[1] += camera_extrinsics[4]; // Y component of camera to calibration point
+		p[2] += camera_extrinsics[5]; // Z component of camera to calibration point
+
+		// The projected pixel coordinates would now be computed. (for now i am not including distortion)
+		T up = p[0] / p[2];
+		T vp = p[1] / p[2];
+
+		// The projected uv pixel values are calculated here based on the current camera intrinsics and extrinsics
+		T projected_u = up * camera_intrinsics[0] + camera_intrinsics[2];
+		T projected_v = vp * camera_intrinsics[1] + camera_intrinsics[3];
+
+		// The residuals are calculated here
+		residuals[0] = projected_u - T(observed_pixel_points_[0]);
+		residuals[1] = projected_v - T(observed_pixel_points_[1]);
+
+		return true;
+	}
+
+	// Factory to hide the construction of the CostFunction object from the client code.
+	static ceres::CostFunction *Create(const double *pixel_points, const double *board_T_point)
+	{
+		return (new ceres::AutoDiffCostFunction<ReProjectionResidual, 2, 6, 4>(
+			new ReProjectionResidual(pixel_points, board_T_point)));
+	}
+
+	// Declare struct variables
+	private:
+		double observed_pixel_points_[3];
+		double board_T_point_[4];
```

Comment:
This is not a transform, so don't use `T` notation.  A better name would be `board_point` or `board_corner_point` (corner point represented in the board frame).  After transforming this into the camera frame, it would be `cam_point` `cam_corner_point`.